### PR TITLE
(maint) sync ConfigConcatenation with upstream

### PR DESCRIPTION
**Pull request by KevinCorcoran**
Most of the diff is just moving stuff around.  Take care of the duplicate `initialize` bug.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,5 @@
 source "https://rubygems.org"
 
+gem 'rspec', :require => 'spec'
```
Comment (jpinsonault) [inline Gemfile]: I don't really know anything about ruby packaging, so I'm just curious, why add this to the Gemfile when it's already in `hocon.gemspec`?

**Comment by jpinsonault** [comment]
Not sure if my question is important, otherwise I'm :+1: on this

File: Gemfile (hunk 1 above)
Comment (KevinCorcoran) [inline Gemfile]: Hmm, yeah, I dunno either.  I hit a problem running spec tests but I can't seem to reproduce it anymore, so I'll just remove this commit.

File: lib/hocon/impl/config_concatenation.rb [hunk 2]
```diff
@@ -186,34 +212,34 @@ def resolve_substitutions(context, source)
...
-    pieces.each do |p|
-      if p.is_a?(Hocon::Impl::ConfigConcatenation)
-        raise ConfigBugOrBrokenError, "ConfigConcatenation should never be nested: #{self}"
-      end
-      if p.is_a?(Unmergeable)
-        had_unmergeable = true
-      end
+  def replace_child(child, replacement)
+    new_pieces = replace_child_in_list(@pieces, child, replacement)
+    if new_pieces == nil
```
Comment (fpringvaldsen) [inline lib/hocon/impl/config_concatenation.rb]: Can probably just be `new_pieces.nil?`

File: lib/hocon/impl/config_concatenation.rb [hunk 3]
```diff
@@ -237,4 +264,14 @@ def render_value_to_sb(sb, indent, at_root, options)
...
     end
   end
+
+  private
+
+  def not_resolved
+    ConfigNotResolvedError.new("need to Config#resolve(), see the API docs for Config#resolve(); substitution not resolved: #{self}")
+  end
+
+  def self.is_ignoring_whitespace(value)
```
Comment (fpringvaldsen) [inline lib/hocon/impl/config_concatenation.rb]: I think this should be `is_ignored_whitespace`

**Comment by fpringvaldsen** [comment]
:+1: I just had two super minor, nitpicky comments. @KevinCorcoran Feel free to ignore them if you'd like, I'm good with merging this now if you don't want to address them.

File: lib/hocon/impl/config_concatenation.rb (hunk 2 above)
Comment (KevinCorcoran) [inline lib/hocon/impl/config_concatenation.rb]: definitely can, but I've been writing these conditions like this to match the Java code as closely as possible.

Comment (fpringvaldsen) [inline lib/hocon/impl/config_concatenation.rb]: That's fine then, I'm good with leaving this

File: lib/hocon/impl/config_concatenation.rb (hunk 3 above)
Comment (KevinCorcoran) [inline lib/hocon/impl/config_concatenation.rb]: good catch, will fix.

**Comment by KevinCorcoran** [comment]
@fpringvaldsen - thanks for reviewing.  I just added a commit to fix that method name.

**Comment by fpringvaldsen** [comment]
:+1: Will merge once Travis finishes

**Comment by cprice404** [comment]
I grabbed this locally after merging @jpinsonault 's PR, rebased it onto master (no conflicts), and ran tests.  seemed fine.

**Pull request merged**